barretenberg/cpp/src/barretenberg/ecc/fields/field_conversion.hpp

```diff
@@ -147,65 +87,28 @@ template <typename T> T convert_from_bn254_frs(std::span<const bb::fr> fr_vec)
  * @param val
  * @return std::vector<bb::fr>
  */
-template <std::integral T> std::vector<bb::fr> inline convert_to_bn254_frs(const T& val)
-{
-    std::vector<bb::fr> fr_vec{ val };
-    return fr_vec;
-}
-
-std::vector<bb::fr> convert_to_bn254_frs(const grumpkin::fr& val);
-
-std::vector<bb::fr> convert_to_bn254_frs(const bb::fr& val);
-
-std::vector<bb::fr> convert_to_bn254_frs(const curve::BN254::AffineElement& val);
-
-std::vector<bb::fr> convert_to_bn254_frs(const curve::Grumpkin::AffineElement& val);
-
-template <size_t N> std::vector<bb::fr> inline convert_to_bn254_frs(const std::array<bb::fr, N>& val)
-{
-    std::vector<bb::fr> fr_vec(val.begin(), val.end());
-    return fr_vec;
-}
-
-template <size_t N> std::vector<bb::fr> inline convert_to_bn254_frs(const std::array<grumpkin::fr, N>& val)
-{
-    std::vector<bb::fr> fr_vec;
-    for (size_t i = 0; i < N; ++i) {
-        auto tmp_vec = convert_to_bn254_frs(val[i]);
-        fr_vec.insert(fr_vec.end(), tmp_vec.begin(), tmp_vec.end());
-    }
-    return fr_vec;
-}
-
-template <size_t N> std::vector<bb::fr> inline convert_to_bn254_frs(const bb::Univariate<bb::fr, N>& val)
-{
-    std::vector<bb::fr> fr_vec;
-    for (size_t i = 0; i < N; ++i) {
-        auto tmp_vec = convert_to_bn254_frs(val.evaluations[i]);
-        fr_vec.insert(fr_vec.end(), tmp_vec.begin(), tmp_vec.end());
-    }
-    return fr_vec;
-}
-
-template <size_t N> std::vector<bb::fr> inline convert_to_bn254_frs(const bb::Univariate<grumpkin::fr, N>& val)
-{
-    std::vector<bb::fr> fr_vec;
-    for (size_t i = 0; i < N; ++i) {
-        auto tmp_vec = convert_to_bn254_frs(val.evaluations[i]);
-        fr_vec.insert(fr_vec.end(), tmp_vec.begin(), tmp_vec.end());
-    }
-    return fr_vec;
-}
-
-template <typename AllValues> std::vector<bb::fr> inline convert_to_bn254_frs(const AllValues& val)
-{
-    auto data = val.get_all();
-    std::vector<bb::fr> fr_vec;
-    for (auto& item : data) {
-        auto tmp_vec = convert_to_bn254_frs(item);
-        fr_vec.insert(fr_vec.end(), tmp_vec.begin(), tmp_vec.end());
+template <typename T> std::vector<bb::fr> convert_to_bn254_frs(const T& val)
+{
+    if constexpr (IsAnyOf<T, bool, uint32_t, bb::fr>) {
+        std::vector<bb::fr> fr_vec{ val };
+        return fr_vec;
+    } else if constexpr (IsAnyOf<T, grumpkin::fr>) {
```

Comment:
I still am oldschool so constexpr is a weird way to do this for me, but if the team likes it :)